utils/ixdtf/src/parsers/records.rs

```diff
@@ -0,0 +1,185 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! The records that `ixdtf`'s contain the resulting values of parsing.
+
+use alloc::vec::Vec;
+
+/// An `IxdtfParseRecord` is an intermediary record returned by `IxdtfParser`.
+#[non_exhaustive]
+#[derive(Default, Debug, PartialEq)]
+pub struct IxdtfParseRecord<'a> {
+    /// Parsed `DateRecord`
+    pub date: Option<DateRecord>,
+    /// Parsed `TimeRecord`
+    pub time: Option<TimeRecord>,
+    /// Parsed UtcOffset
+    pub offset: Option<UTCOffsetRecord>,
+    /// Parsed `TimeZone` annotation with critical flag and data (UTCOffset | IANA name)
+    pub tz: Option<TimeZoneAnnotation<'a>>,
+    /// The parsed calendar value.
+    pub calendar: Option<&'a str>,
+    /// A collection of annotations provided on an IXDTF string.
+    pub annotations: Vec<Annotation<'a>>,
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq)]
+/// A record of an annotation.
+pub struct Annotation<'a> {
+    /// Whether this annotation is flagged as critical
+    pub critical: bool,
+    /// The parsed key value of the annotation
+    pub key: &'a str,
+    /// The parsed value of the annotation
+    pub value: &'a str,
+}
+
+#[allow(clippy::exhaustive_structs)] // DateRecord only allows for a year, month, and day value.
+#[derive(Default, Debug, Clone, Copy, PartialEq)]
+/// The record of a parsed date.
+pub struct DateRecord {
+    /// Date Year
+    pub year: i32,
+    /// Date Month
+    pub month: u8,
+    /// Date Day
+    pub day: u8,
+}
+
+/// Parsed Time info
+#[allow(clippy::exhaustive_structs)] // TimeRecord only allows for a hour, minute, second, and sub-second value.
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+pub struct TimeRecord {
+    /// An hour
+    pub hour: u8,
+    /// A minute value
+    pub minute: u8,
+    /// A second value.
+    pub second: u8,
+    /// A nanosecond value representing all sub-second components.
+    pub nanosecond: u32,
+}
+
+/// A `TimeZoneAnnotation` that represents a parsed `TimeZoneRecord` and its critical flag.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq)]
+pub struct TimeZoneAnnotation<'a> {
+    /// Critical flag for the `TimeZoneAnnotation`.
+    pub critical: bool,
+    /// The parsed `TimeZoneRecord` for the annotation.
+    pub tz: TimeZoneRecord<'a>,
+}
+
+/// Parsed `TimeZone` data, which can be either a UTC Offset value or IANA Time Zone Name value.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq)]
+pub enum TimeZoneRecord<'a> {
+    /// TimeZoneIANAName
+    Name(&'a str),
+    /// TimeZoneOffset
+    Offset(UTCOffsetRecord),
+}
+
+/// The parsed sign value, representing whether its struct is positive or negative.
+#[repr(i8)]
+#[allow(clippy::exhaustive_enums)] // Sign can only be positive or negative.
```

Comment:
Question: there are often 3 types of signs, known as `Signum`, which are "explicit minus", "explicit plus", and "no sign", as in "-1", "+1" and "1". Does IXDTF make a distinction?

Reply:
IXDTF does not to my knowledge and from what I can find. RFC3339 does layout that the offset must be "+" or "-". So Duration is the only one that can be "no sign" as it is considered optional.